modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/rebalance/DistributionZoneRebalanceEngine.java

```diff
@@ -0,0 +1,286 @@
+    }
+
+    private WatchListener createDistributionZonesDataNodesListener() {
+        return new WatchListener() {
+            @Override
+            public CompletableFuture<Void> onUpdate(WatchEvent evt) {
+                if (!busyLock.enterBusy()) {
+                    return failedFuture(new NodeStoppingException());
+                }
+
+                try {
+                    byte[] dataNodesBytes = evt.entryEvent().newEntry().value();
+
+                    if (dataNodesBytes == null) {
+                        //The zone was removed so data nodes was removed too.
+                        return completedFuture(null);
+                    }
+
+                    NamedConfigurationTree<TableConfiguration, TableView, TableChange> tables = tablesConfiguration.tables();
+
+                    int zoneId = extractZoneId(evt.entryEvent().newEntry().key());
+
+                    Set<String> dataNodes = dataNodes(ByteUtils.fromBytes(dataNodesBytes));
+
+                    for (int i = 0; i < tables.value().size(); i++) {
+                        TableView tableView = tables.value().get(i);
+
+                        int tableZoneId = tableView.zoneId();
+
+                        DistributionZoneConfiguration distributionZoneConfiguration =
+                                getZoneById(zonesConfiguration, tableZoneId);
+
+                        if (zoneId == tableZoneId) {
+                            TableConfiguration tableCfg = tables.get(tableView.name());
+
+                            byte[] assignmentsBytes = ((ExtendedTableConfiguration) tableCfg).assignments().value();
+
+                            List<Set<Assignment>> tableAssignments = assignmentsBytes == null ? Collections.emptyList() :
+                                    ByteUtils.fromBytes(assignmentsBytes);
+
+                            for (int part = 0; part < distributionZoneConfiguration.partitions().value(); part++) {
+                                UUID tableId = ((ExtendedTableConfiguration) tableCfg).id().value();
+
+                                TablePartitionId replicaGrpId = new TablePartitionId(tableId, part);
+
+                                int replicas = distributionZoneConfiguration.replicas().value();
+
+                                int partId = part;
+
+                                updatePendingAssignmentsKeys(
+                                        tableView.name(),
+                                        replicaGrpId,
+                                        dataNodes,
+                                        replicas,
+                                        evt.entryEvent().newEntry().revision(),
+                                        metaStorageManager,
+                                        part,
+                                        tableAssignments.isEmpty() ? Collections.emptySet() : tableAssignments.get(part)
+                                ).exceptionally(e -> {
```

Comment:
Why is `tableAssignments.isEmpty()` done here and not done in onUpdateReplicas?

Reply:
Because it's harmless test related code that makes it possible to use table configuration injection in DistributionZoneRebalanceEngineTest
```
            @InjectConfiguration
                    ("mock.tables {"
                            + "table0 = { zoneId = 1 },"
                            + "table1 = { zoneId = 1 },"
                            + "table2 = { zoneId = 2 },"
                            + "table3 = { zoneId = 2 },"
                            + "table4 = { zoneId = 2 },"
                            + "table5 = { zoneId = 2 }}")
            TablesConfiguration tablesConfiguration
```
Otherwise test will fail because of null assignments.

Reply:
Ok